### ISSUE TITLE
Integrating with external parsers of type `Text -> Either String a` or similar

### ISSUE THREAD

**Issue by michalrus** [dhall-lang/dhall-haskell#609]
Hey!

I wanted to define an `instance Interpret UUID`, so, looking at the types, I did this:

```haskell
instance Interpret UUID where
  autoWith o =
    let Type a b = autoWith o
     in Type (UUID.fromText <=< a) b
```

And the happy path works nicely, but if the user supplies an incorrectly formatted Text, we’re getting:

```
ping-node: Error: Invalid Dhall.Type

Every Type must provide an extract function that succeeds if an expression
matches the expected type.  You provided a Type that disobeys this contract
```

How to go about this? Thank you! =)

**Comment by Profpatsch** [comment]
There really is a fixed set of types that dhall supports.

I tried doing something like this myself, but it’s impossible to achieve this way (exactly because you can’t implement a sensible version of `Type`’s `expected`, since there is no dhall `Expr` literal for your custom type).

See https://github.com/dhall-lang/dhall-lang/wiki/How-to-add-a-new-built-in-function for a tutorial on adding custom symbols.
I think there was also a tutorial describing how to achieve it by something like

```
\(UUID: Type) ->
\(Text/toUUID: Text -> UUID) ->
…
```

and then supplying the arguments from Haskell, but I can’t find it.

**Comment by Gabriel439** [comment]
@michalrus: This issue looks very similar to https://github.com/dhall-lang/dhall-haskell/issues/387, which is about providing a way to customize the error message returned by Dhall when `extract` fails

**Comment by michalrus** [comment]
@Profpatsch I seeee, mhmmm, thank you.

@Gabriel439 But there you write:

> Yeah, if it's a simple change. Keep in mind that if a `Type` is implemented correctly then `extract` should never fail, so I wouldn't go overboard in improving the error message here

However, this issue seems a bit different, then, right? I can’t make `expect` expect `Text`s only in UUID format. So a detailed error message (including path), as @ocharles suggests, would be beneficial in this case. Or is the reasoning flawed?

**Comment by Gabriel439** [comment]
@michalrus: I would be fine allowing the user to customize the error message via the API

**Comment by ocharles** [comment]
If we're going to allow `extract` to return a failure error message, then I think we're heading into a territory where 

> Every `Type` must obey the contract that if an expression's type matches the the `expected` type then the `extract` function must succeed. 

no longer really holds. Should this be weakened? If not, then it's simply the case that `Interpret UUID` can't exist. This doesn't mean you can't parse UUIDs, but that `Interpret` is not the right entry point, and that also means giving up the current generics support.

**Comment by Gabriel439** [comment]
@ocharles: So the part where the contract matters most is in the `Interpret` instance for functions, which uses a partial pattern match when attempting to decode the function's result:

https://github.com/dhall-lang/dhall-haskell/blob/6fd4c8372463c2998580c0e8be38b547e4038861/src/Dhall.hs#L803

So if we don't want to risk that then we need to come up with another mechanism for dealing with decoding faliures

**Comment by Gabriel439** [comment]
Closing as I believe this was fixed by #1011
